addons/mass_mailing/models/mailing.py

```diff
@@ -1306,36 +1315,117 @@ def _unsubscribe_token(self, res_id, email):
 
-    def _convert_inline_images_to_urls(self, body_html):
-        """
-        Find inline base64 encoded images, make an attachement out of
-        them and replace the inline image with an url to the attachement.
-        """
 
-        def _image_to_url(b64image: bytes):
-            """Store an image in an attachement and returns an url"""
-            attachment = self.env['ir.attachment'].create({
-                'datas': b64image,
-                'name': "cropped_image_mailing_{}".format(self.id),
-                'type': 'binary',})
+    def _image_to_url(self, b64image: bytes):
+        """Store an image in an attachement and returns an url"""
+        attachment = self.env['ir.attachment'].create({
+            'datas': b64image,
+            'name': "cropped_image_mailing_{}".format(self.id),
+            'type': 'binary',})
+
+        attachment.generate_access_token()
 
-            attachment.generate_access_token()
+        return '/web/image/%s?access_token=%s' % (
+            attachment.id, attachment.access_token)
 
-            return '/web/image/%s?access_token=%s' % (
-                attachment.id, attachment.access_token)
+    def _get_image_by_url(self, url, session):
+        maxsize = int(tools.config.get("import_image_maxbytes", DEFAULT_IMAGE_MAXBYTES))
+        _logger.debug("Trying to import image from URL: %s", url)
+        try:
+            response = session.get(url, timeout=int(tools.config.get("import_image_timeout", DEFAULT_IMAGE_TIMEOUT)))
+            response.raise_for_status()
+
+            if response.headers.get('Content-Length') and int(response.headers['Content-Length']) > maxsize:
+                raise ImportValidationError(
+                    _("File size exceeds configured maximum (%s bytes)", maxsize)
+                )
+
+            content = bytearray()
+            for chunk in response.iter_content(DEFAULT_IMAGE_CHUNK_SIZE):
+                content += chunk
+                if len(content) > maxsize:
+                    raise ImportValidationError(
+                        _("File size exceeds configured maximum (%s bytes)", maxsize)
+                    )
+
+            image = Image.open(io.BytesIO(content))
+            w, h = image.size
+            if w * h > 42e6:  # Nokia Lumia 1020 photo resolution
+                raise ImportValidationError(
+                    _("Image size excessive, imported images must be smaller than 42 million pixel")
+                )
 
-        modified = False
-        root = lxml.html.fromstring(body_html)
+            return content
+        except Exception as e:
+            _logger.exception(e)
+            raise ImportValidationError(
+                _(
+                    "Could not retrieve URL: %(url)s: %(error)s",
+                    url=url, error=e
+                )
+            )
+
+    def _convert_inline_images_to_urls(self, root):
```

Comment:
You should not change the type of parameters in stable ... you should keep html (and keep the same variable name). You could maybe do some guess-based computation (like: if it is str -> parse it).

Reply:
```diff
diff --git a/addons/mass_mailing/models/mailing.py b/addons/mass_mailing/models/mailing.py
index 4a099b2c937..4c17356dc7e 100644
--- a/addons/mass_mailing/models/mailing.py
+++ b/addons/mass_mailing/models/mailing.py
@@ -1365,12 +1365,13 @@ class MassMailing(models.Model):
                 )
             )
 
-    def _convert_inline_images_to_urls(self, root):
+    def _convert_inline_images_to_urls(self, body_html):
         """
         Find inline base64 encoded images, make an attachement out of
         them and replace the inline image with an url to the attachement.
         """
         did_convert = False
+        root = lxml.html.fromstring(body_html) if isinstance(body_html, str) else body_html
         for node in root.iter('img'):
             match = image_re.match(node.attrib.get('src', ''))
             if match:
```

Reply:
Done here: https://github.com/odoo/odoo/pull/119325/commits/c7f282c8d6cab461ebb8c1507540444b522c597a